smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddHttpApiKeyAuthPlugin.java

```diff
@@ -0,0 +1,169 @@
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.ServiceIndex;
+import software.amazon.smithy.model.knowledge.TopDownIndex;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.traits.HttpApiKeyAuthTrait;
+import software.amazon.smithy.model.traits.OptionalAuthTrait;
+import software.amazon.smithy.typescript.codegen.CodegenUtils;
+import software.amazon.smithy.typescript.codegen.TypeScriptSettings;
+import software.amazon.smithy.typescript.codegen.TypeScriptWriter;
+import software.amazon.smithy.utils.IoUtils;
+import software.amazon.smithy.utils.ListUtils;
+import software.amazon.smithy.utils.SmithyInternalApi;
+
+/**
+ * Add config and middleware to support a service with the @httpApiKeyAuth trait.
+ */
+@SmithyInternalApi
+public final class AddHttpApiKeyAuthPlugin implements TypeScriptIntegration {
+
+    /**
+     * Plug into code generation for the client.
+     *
+     * This adds the configuration items to the client config and plugs in the
+     * middleware to operations that need it.
+     *
+     * The middleware will inject the client's configured API key into the
+     * request as defined by the @httpApiKeyAuth trait. If the trait says to
+     * put the API key into a named header, that header will be used, optionally
+     * prefixed with a scheme. If the trait says to put the API key into a named
+     * query parameter, that query parameter will be used.
+     */
+    @Override
+    public List<RuntimeClientPlugin> getClientPlugins() {
+        return ListUtils.of(
+            // Add the config if the service uses HTTP API key authorization.
+            RuntimeClientPlugin.builder()
+                    .inputConfig(Symbol.builder()
+                            .namespace("./" + CodegenUtils.SOURCE_FOLDER + "/middleware/HttpApiKeyAuth", "/")
+                            .name("HttpApiKeyAuthInputConfig")
+                            .build())
+                    .resolvedConfig(Symbol.builder()
+                            .namespace("./" + CodegenUtils.SOURCE_FOLDER + "/middleware/HttpApiKeyAuth", "/")
+                            .name("HttpApiKeyAuthResolvedConfig")
+                            .build())
+                    .resolveFunction(Symbol.builder()
+                            .namespace("./" + CodegenUtils.SOURCE_FOLDER + "/middleware/HttpApiKeyAuth", "/")
+                            .name("resolveHttpApiKeyAuthConfig")
+                            .build())
+                    .servicePredicate((m, s) -> hasEffectiveHttpApiKeyAuthTrait(m, s))
+                    .build(),
+
+            // Add the middleware to operations that use HTTP API key authorization.
+            RuntimeClientPlugin.builder()
+                    .pluginFunction(Symbol.builder()
+                            .namespace("./" + CodegenUtils.SOURCE_FOLDER + "/middleware/HttpApiKeyAuth", "/")
```

Comment:
Refs: https://github.com/awslabs/smithy-typescript/pull/473#discussion_r774711588
```suggestion
            // Add the middleware to operations that use HTTP API key authorization.
            RuntimeClientPlugin.builder()
                    .pluginFunction(Symbol.builder()
                            .namespace("./" + CodegenUtils.SOURCE_FOLDER + "/middleware/HttpApiKeyAuth", "/")
```

Reply:
Add `apiKey` config by overriding `addConfigInterfaceFields` instead.

Example https://github.com/awslabs/smithy-typescript/blob/d263078b81485a6a2013d243639c0c680343ff47/smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddChecksumRequiredDependency.java#L45-L69

Reply:
Tried this, see resulting error message when trying to compile the generated client. https://github.com/awslabs/smithy-typescript/pull/473#discussion_r775072233